index.js

```diff
@@ -153,54 +153,87 @@ function SpotifyWebHelper(opts) {
 			});
 		});
 	};
-	this.generateSpotifyUrl = function (url) {
+	this.generateSpotifyUrl = function (url, port) {
+		port = typeof port === "undefined" ? localPort : port;
 		var protocol = 'https://';
-		if (localPort >= START_HTTP_PORT && localPort <= END_HTTP_PORT) {
+		if (port >= START_HTTP_PORT && port <= END_HTTP_PORT) {
 			protocol = 'http://';
 		}
-		return util.format('%s%s:%d%s', protocol, '127.0.0.1', localPort, url);
+		return util.format('%s%s:%d%s', protocol, '127.0.0.1', port, url);
 	};
 	this.getOauthToken = function () {
 		return new Promise(function (resolve, reject) {
 			getJSON({
 				url: 'http://open.spotify.com/token'
 			})
-			.then(function (res) {
-				resolve(res.t);
-			})
-			.catch(reject);
+				.then(function (res) {
+					resolve(res.t);
+				})
+				.catch(reject);
 		});
 	};
-  this.checkForError = status => {
-    if (!status.open_graph_state) {
-      this.player.emit('error', new Error('No user logged in'));
-      return true;
-    }
-    if (status.error) {
-      this.player.emit('error', new Error(status.error.message));
-      return true;
-    }
-    return false;
-  };
-	this.detectPort = function () {
+	this.checkForError = status => {
+		if (!status.open_graph_state) {
+			this.player.emit('error', new Error('No user logged in'));
+			return true;
+		}
+		if (status.error) {
+			this.player.emit('error', new Error(status.error.message));
+			return true;
+		}
+		return false;
+	};
+	this.checkPort = function(port) {
 		return getJSON({
-			url: this.generateSpotifyUrl('/service/version.json'),
+			url: this.generateSpotifyUrl('/service/version.json', port),
 			headers: ORIGIN_HEADER,
 			params: {
 				'service': 'remote'
 			}
-		})
-		.then(() => localPort)
-		.catch((err) => {
-			if (localPort === END_HTTP_PORT) {
+		}).then(() => {
+				return localPort;
+			})
+			.catch((err) => {
 				throw err;
-			} else if (localPort === END_HTTPS_PORT) {
-				localPort = START_HTTP_PORT;
-			} else {
-				++localPort;
-			}
-			return this.detectPort();
-		});
+			});
+	};
+	this.portCheckCount = 0;
+	this.foundPort = false;
+	this.checkPorts = function(start,end){
+		//console.log(start,end);
+		for(let i = start; i <= end; i++) {
```

Comment:
If you really care about speed you might be interested in [Promise.race](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/race).  This can simplify this process a lot, you just need an array of each port being checked rather than a loop.

Reply:
I have bookmarked that, thats really useful tbh. If it wasnt for the fact that this was just a quick fix for something i was testing i would have probably tried to cancel the rest of the requests though i didnt know about the race. Does the race part work in browser?

Reply:
Promise.race works on all platforms that support promises.

Reply:
Yea i quickly checked after on the caniuse website and it seems only opera mini doesnt support it so its fine for use basically